### PR TITLE
Establish workflow framework, and add the ungrib task 

### PR DESCRIPTION
**Pull request by guoqing-noaa**
This PR addresses issue #435 

The whole workflow framework is committed but we only turn on the `ungrib `task at the moment.
We will incrementally add other tasks in the subsequent PRs.

The build_and_run.md is updated in this PR and can be accessed through this quick link:
https://github.com/guoqing-noaa/rrfs-workflow/blob/ungrib/doc/build_and_run.md

A quick guide for testing:
```
git clone -b ungrib --recursive git@github.com:guoqing-noaa/rrfs-workflow
cd sorc
./build.all
cd ../rocoto
cp ../parm/config/exp/exp.setup .
vi exp.setup # modfiy OPSROOT to your own path
./setup_exp.py
```

Go to `EXPDIR`, open `rrfs.xml`, and make sure it has all the required tasks and settings.
Use `./run_rocoto.sh` to run the experiment. 

**This PR was contributed by: Guoqing Ge, Ming Hu, Chunhua Zhou, Sijie Pan, Eric Jame, and Junjun Hu**

### CODE REVIEW

**Review by ShunLiu-NOAA — APPROVED** [review]
looks good to me.

File: scripts/exrrfs_ungrib.sh
Comment (MatthewPyle-NOAA) [inline scripts/exrrfs_ungrib.sh:1]: WGRIB2 shouldn't be defined this way down at the scripts level.  Ideally would be defined by a module load in the submitting job (the .ecf submission job in production)

**Review by MatthewPyle-NOAA — COMMENTED** [review]
Mostly looks good to me - but need some cleanup on how WGRIB2 is defined.

**Comment by ShunLiu-NOAA** [comment]
@guoqing-noaa could you please take a look at RAP/HRRR ungrib script?

**Comment by lgannoaa** [comment]
Thank you @guoqing-noaa for your hard work on it.
If this PR supposed to only be focus on what described on its title - add the Ungrib task. Would it be better if we can separate into groups of small PR on these 421 files? Looks to me, maybe multiple PR referenced to the github issue number will be a good way to handle it.

**Comment by guoqing-noaa** [comment]
FYI, Shun, Lin, and I had a meeting today and it is preferred to move forward with the current form for this PR.
Feel free to chime in if you have any concerns. Thanks!

**Comment by TingLei-NOAA** [comment]
@guoqing-noaa  Is my understanding right  that this PR is to add new scripts and not changing  any existing workflow scripts? And, so, any users who don't need to use the new functions for being now will not be affected?

**Comment by guoqing-noaa** [comment]
> @guoqing-noaa Is my understanding right that this PR is to add new scripts and not changing any existing workflow scripts? And, so, any users who don't need to use the new functions for being now will not be affected?

@TingLei-NOAA Please see the previous PR #434 for more information.

**Comment by guoqing-noaa** [comment]
> @guoqing-noaa The ungrib changes look good to me. I am unfamiliar with most of the MPAS components, so I didn't pay as much attention to those, but I'm more familiar with ungrib since it is also part of WRF.
> 
> One question I had is related to exrrfs_ungrib.sh. Why is cpreq defined as a variable on line 4? I would think you could simply use "cpreq" further down in the script without needing to define it, but maybe there is a good reason for doing this. Thanks!

@BenjaminBlake-NOAA Thanks a lot for your approval!  Because there are lots of back-and-forth coping (stmp -> com -> stmp) in the current setup, so redefine cpreq to 'ln -snf' will benefit the develop work. Now that @lgannoaa got clarifications from NCO on how to share data inside and between cycles and greatly reduce the duplicate copying work, I will try to use cpreq directly.

**Comment by guoqing-noaa** [comment]
@MatthewPyle-NOAA @ShunLiu-NOAA @lgannoaa @BenjaminBlake-NOAA 

FYI, we have had a few meetings as well as email communications with Lin and Shun. Lin also reached out to the NCO and got more clarifications. Based on the latest situation, there is a desire to close this PR and we will do the following to move forward:

1. split this PR into a series of small PRs to facilitate the review process
2. use the umbrella DATA  directories to share data between tasks inside the same cycle
3. use the com directory to share data between cycles
4. rename rocoto/ to workflow/ which will have `setup_rocoto.py` and `setup_ecf.py` for setting up rocoto workflow and ecflow respectively. Relevant Pyhont functions will be put under workflow/ecf_funcs and workflow/rocoto_funcs respectively.

Let me know if you have any questions on this. Thanks!

**Comment by lgannoaa** [comment]
Thank you, @guoqing-noaa for your work. I agree with close this PR for now.

**Comment by TingLei-NOAA** [comment]
> > @guoqing-noaa Is my understanding right that this PR is to add new scripts and not changing any existing workflow scripts? And, so, any users who don't need to use the new functions for being now will not be affected?
> 
> @TingLei-NOAA Please see the previous PR #434 for more information.

@guoqing-noaa  Can I take your answer including your description in PR#434 as a "Yes" to my question?

**Comment by guoqing-noaa** [comment]
I have received many requests from developers at GSL and EMC. They would like to have this PR to get merged as soon as possible because their project work is heavily dependent on this PR. So as I mentioned in a previous comment, it is preferred to merge this PR in its current form and we can enhance/improve it gradually. 

There are 421 file changes in this PR, but most of them don't need a review at the PR level:

361 (minus the number of subdirectories) of them are fix files, which do not need a review at this PR level. We have a different fix file management group to handle this. It is safe to ignore this part.
7 are modulefiles which does not need a review for operational use.
8 are pythons files which can be used to set up a rocoto workflow that does not need a review for operational use
11 are utils/, which do not need a review for operational use
12 are sorc/ files which does not need a review at the workflow level
34 are parm files that have been modified according to review comments.

So the only parts that need reviewing are jobs/ and scripts/ which have been approved by two developers.

**Comment by ShunLiu-NOAA** [comment]
@MatthewPyle-NOAA Could you please review this PR? We can open ISSUE to continue discussing the potential issues and address them in the future PR. We should be able to fix the "nwges" path easily by adding configuration option when we generate workflow, however, we can address this in another PR.

**Review by TingLei-NOAA — APPROVED** [review]
Since this PR will not affect current functions, I would approve this PR because it will give RRFS developers a good starting point to run and develop related tests and functions. 
Further improvement (even fixes)  could be done when it is used through this repo by RRFS developers.

**Review by MatthewPyle-NOAA — APPROVED** [review]
While ideally this would have been handled in some smaller, easier to review chunks, I don't want to stand in the way of this initial step forward.

**Comment by guoqing-noaa** [comment]
Thanks, @MatthewPyle-NOAA !

**Pull request merged**